File: cpp_library/testrobot/src/actions.cpp
#include <actions.hpp>
#include <main.hpp>

#include <arpirobot/core/log/Logger.hpp>
#include <arpirobot/core/network/NetworkTable.hpp>


void TestAction::begin(){
    pid.reset();
    pid.setMax(1.0);
    pid.setMin(-1.0);
    pid.setKp(1.0);
    pid.setKi(0.0);
    pid.setKd(0.0);
    pid.setSetpoint(Main::robot->imu.getGyroZ() + 90.0);
}

void TestAction::process(){
    if(NetworkTable::changed("P")){
        try{
            pid.setKp(std::stod(NetworkTable::get("P")));
        }catch(const std::invalid_argument &e){
            // Do not change P
        }
    }
    if(NetworkTable::changed("I")){
        try{
            pid.setKi(std::stod(NetworkTable::get("I")));
        }catch(const std::invalid_argument &e){
            // Do not change I
        }
    }
    if(NetworkTable::changed("D")){
        try{
            pid.setKd(std::stod(NetworkTable::get("D")));
        }catch(const std::invalid_argument &e){
            // Do not change D
        }
    }
    Main::robot->driveHelper.updateRotation(-pid.getOutput(Main::robot->imu.getGyroZ()));
}

void TestAction::finish(bool wasInterrupted){
    Main::robot->driveHelper.updateRotation(0);
}

bool TestAction::shouldContinue(){
    return true;
}

File: testrobot_templates/py/actions.py

from arpirobot.core.audio import AudioManager
from arpirobot.core.log import Logger
from arpirobot.core.action import Action
from arpirobot.core.control import PID
from arpirobot.core.network import NetworkTable

import main 

class TestAction(Action):
    def __init__(self):
        super().__init__()
        self.pid = PID()

    def begin(self):
        self.pid.reset()
        self.pid.set_max(1.0)
        self.pid.set_min(-1.0)
        self.pid.set_kp(1.0)
        self.pid.set_ki(0.0)
        self.pid.set_kd(0.0)
        self.pid.set_setpoint(main.robot.imu.get_gyro_z() + 90.0)

    def process(self):
        if NetworkTable.changed("P"):
            try:
                self.pid.set_kp(float(NetworkTable.get("P")))
            except ValueError:
                pass
        if NetworkTable.changed("I"):
            try:
                self.pid.set_ki(float(NetworkTable.get("I")))
            except ValueError:
                pass
        if NetworkTable.changed("D"):
            try:
                self.pid.set_kd(float(NetworkTable.get("D")))
            except ValueError:
                pass
        main.robot.drive_helper.update_rotation(-self.pid.get_output(main.robot.imu.get_gyro_z()))
    
    def finish(self, was_interrupted: bool):
        main.robot.drive_helper.update_rotation(0)
    
    def should_continue(self) -> bool:
        return True